pkg/xds/bootstrap/template.go

```diff
@@ -77,6 +79,21 @@ static_resources:
               socket_address:
                 address: {{ .XdsHost }}
                 port_value: {{ .XdsPort }}
+{{ if .CertBytes }}
+    transport_socket:
+      name: envoy.transport_sockets.tls
+      typed_config:
+        "@type": type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext
+        sni: {{ .XdsHost }}
+        common_tls_context:
+          tls_params:
+            tls_minimum_protocol_version: TLSv1_2
+          tls_certificates:
+          - certificate_chain:
+              inline_bytes: "{{ .CertBytes }}"
+            private_key:
+              inline_bytes: "{{ .KeyBytes }}"
+{{ end }}
```

Comment:
Hey, can you help me understand this part a bit?
So we don't setup XDS server with TLS, but we setup Envoy with TLS client. Additionally, the key travels through bootstrap.
Can we apply to XDS the same solution that we use for SDS? Meaning, you expose XDS Server in Kuma with TLS. In bootstrap you pass only the cert of the XDS server and in Contour you use passthrough?

Reply:
Hi @jakubdyszkiewicz,
Sure, TLS for XDS is terminated at the reverse proxy. The reverse proxy then connects to the kuma-cp XDS server without TLS. This is the reason i have not added TLS support to XDS. Also the bigger goal is to follow the envoy sidecar pattern with Kuma-cp in case we want to terminate TLS closer to the control plane. 

envoy(XDS) =======| (TLS) reverse proxy (no TLS)| ------------- Kuma-cp(XDS)
bootstrap is secured between the reverse proxy and kuma-dp, but passing the key may not a good idea??
I need to understand the SDS pattern better. Do you have a envoyproxy link which explains this pattern?

Reply:
It's not a pattern in Envoy itself rather in Kuma.

SDS:
When Control Plane (CP) starts we can either provide Cert + Key for SDS server or CP will generate it for us.
Then SDS server is started with TLS-only using this cert + key. When CP configures the Envoy with mTLS between dataplanes it has to provide (through a XDS) a configuration to SDS. With this configuration CP sends only a cert (without a key) to Envoy. Envoy then can establish a TLS connection to the SDS server using this cert.

So with securing the XDS it would look like this:
we can optionally configure XDS with the cert and key. When a user does it, we start a TLS-only XDS server and when generating a bootstrap configuration we configure ADS cluster with TLS certificate (without the key).

This way the connection is secured. If you use a passthrough functionality in Contour as far I understand it does not terminate a TLS but rather just pass a request to the XDS/SDS server which would then terminate it.

Reply:
Hi @jakubdyszkiewicz,
I took your suggestion and implemented the same but I am not able to make progress. I want to validate if my understanding is the same as yours. 
1. I am successful in configuring XDS with tls and i could reach it via passthrough. This is without making the bootstrap changes.
2. For bootstrap template following change. Can you ack if this is what you recommended. 
```
common_tls_context: 
 tls_certificate_sds_secret_configs: 
 - name: default
   sds_config: 
    api_config_source: 
     api_type: GRPC
     grpc_services: 
      - google_grpc: 
        target_uri: "<sds url>"
        channel_credentials: 
         ssl_credentials: 
          root_certs: 
           inline_bytes: "sds cert base 64"
        call_credentials: 
        - from_plugin: 
           name: "envoy.grpc_credentials.file_based_metadata"
           typed_config: 
            "@type": "type.googleapis.com/envoy.config.grpc_credential.v2alpha.FileBasedMetadataConfig"
            secret_data: 
             filename: "<token file>"
        stat_prefix: sdsstat
        credentials_factory_name: "envoy.grpc_credentials.file_based_metadata"
```
this is posing the following challenge:

a.  typed_config is mostly a part of dynamic cluster and hence when kuma-dp receives it its not able to convert the definition to envoy bootstrap definition. https://github.com/envoyproxy/go-control-plane/blob/master/envoy/config/bootstrap/v2/bootstrap.pb.go#L35 
b. if i remove call_credential section, sds stream does not go through. I receive this error: 
"could not parse token: token contains an invalid number of segments"
 
Any suggestions?

Reply:
Hi,

for XDS I would skip the `call_credentials` for now. We can add it later if needed. In my opinion, there is no sensitive information about XDS config. What we care here is that we verify that the TLS cert is from the CP so the XDS config is not provided by impersonated CP.

Reply:
Hi @jakubdyszkiewicz 
I have removed the SDS dependency and added trusted ca as check for XDS. The PR is up for review. 
Also i see that somehow the CI did not trigger with the commit.

Reply:
I merged the latest code and now the CI passed

Reply:
Nice, this is exactly what I had in mind :)